File: tools/export_screenshots.js
#!/usr/bin/env node
/**
 * Resize iPhone screenshots to App Store Connect required dimensions.
 * Letterboxes with black padding; no overlays or text.
 */

import { promises as fs } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import sharp from "sharp";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, "..");
const RAW_DIR = path.join(REPO_ROOT, "appstore", "raw");
const EXPORT_DIR = path.join(REPO_ROOT, "appstore", "export");

/** App Store Connect iPhone screenshot sizes (portrait). */
const SIZES = [
  { width: 1290, height: 2796, label: "1290x2796" },
  { width: 1284, height: 2778, label: "1284x2778" },
  { width: 1242, height: 2688, label: "1242x2688" },
];

const IMAGE_EXTENSIONS = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".webp",
  ".heic",
  ".heif",
]);

async function listSourceImages(dir) {
  let entries;
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch (err) {
    if (err.code === "ENOENT") {
      throw new Error(
        `Source folder missing: ${dir}\nCreate it and add screenshots, then run again.`,
      );
    }
    throw err;
  }

  return entries
    .filter((e) => e.isFile())
    .map((e) => e.name)
    .filter((name) => IMAGE_EXTENSIONS.has(path.extname(name).toLowerCase()))
    .sort((a, b) => a.localeCompare(b, undefined, { numeric: true }));
}

async function exportOne(inputPath, outputPath, width, height) {
  await sharp(inputPath)
    .rotate()
    .resize(width, height, {
      fit: "contain",
      background: { r: 0, g: 0, b: 0, alpha: 1 },
      kernel: sharp.kernel.lanczos3,
    })
    .png({ compressionLevel: 9, adaptiveFiltering: true })
    .toFile(outputPath);
}

async function main() {
  const sources = await listSourceImages(RAW_DIR);
  if (sources.length === 0) {
    console.error(`No images found in ${RAW_DIR}`);
    console.error("Supported: .png .jpg .jpeg .webp .heic .heif");
    process.exit(1);
  }

  await fs.mkdir(EXPORT_DIR, { recursive: true });
  for (const { label } of SIZES) {
    await fs.mkdir(path.join(EXPORT_DIR, label), { recursive: true });
  }

  let exported = 0;
  for (const filename of sources) {
    const inputPath = path.join(RAW_DIR, filename);
    const base = path.parse(filename).name;

    for (const { width, height, label } of SIZES) {
      const outputPath = path.join(EXPORT_DIR, label, `${base}.png`);
      await exportOne(inputPath, outputPath, width, height);
      exported += 1;
      console.log(`✓ ${label}/${base}.png`);
    }
  }

  console.log(
    `\nDone: ${sources.length} source(s) → ${exported} file(s) in ${EXPORT_DIR}`,
  );
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: tools/export_screenshots.py
#!/usr/bin/env python3
"""
Resize iPhone screenshots to App Store Connect required dimensions.
Letterboxes with black padding; no overlays or text.
"""

from __future__ import annotations

import sys
from pathlib import Path

try:
    from PIL import Image, ImageOps
except ImportError:
    print("Pillow is required: pip install -r requirements.txt", file=sys.stderr)
    sys.exit(1)

REPO_ROOT = Path(__file__).resolve().parent.parent
RAW_DIR = REPO_ROOT / "appstore" / "raw"
EXPORT_DIR = REPO_ROOT / "appstore" / "export"

SIZES: list[tuple[int, int, str]] = [
    (1290, 2796, "1290x2796"),
    (1284, 2778, "1284x2778"),
    (1242, 2688, "1242x2688"),
]

IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg", ".webp", ".heic", ".heif"}


def list_source_images() -> list[Path]:
    if not RAW_DIR.is_dir():
        raise SystemExit(
            f"Source folder missing: {RAW_DIR}\n"
            "Create it and add screenshots, then run again."
        )

    files = [
        p
        for p in RAW_DIR.iterdir()
        if p.is_file() and p.suffix.lower() in IMAGE_EXTENSIONS
    ]
    return sorted(files, key=lambda p: p.name)


def export_one(src: Path, dest: Path, width: int, height: int) -> None:
    with Image.open(src) as im:
        im = ImageOps.exif_transpose(im)
        im = im.convert("RGBA")
        fitted = ImageOps.contain(im, (width, height), method=Image.Resampling.LANCZOS)
        canvas = Image.new("RGB", (width, height), (0, 0, 0))
        x = (width - fitted.width) // 2
        y = (height - fitted.height) // 2
        canvas.paste(fitted, (x, y), fitted)
        dest.parent.mkdir(parents=True, exist_ok=True)
        canvas.save(dest, format="PNG", optimize=True)


def main() -> None:
    sources = list_source_images()
    if not sources:
        print(f"No images found in {RAW_DIR}", file=sys.stderr)
        print("Supported: .png .jpg .jpeg .webp .heic .heif", file=sys.stderr)
        sys.exit(1)

    exported = 0
    for src in sources:
        base = src.stem
        for width, height, label in SIZES:
            dest = EXPORT_DIR / label / f"{base}.png"
            export_one(src, dest, width, height)
            exported += 1
            print(f"✓ {label}/{base}.png")

    print(
        f"\nDone: {len(sources)} source(s) → {exported} file(s) in {EXPORT_DIR}"
    )


if __name__ == "__main__":
    main()
